### PR TITLE
Add remote cache and CI configs

### PR DESCRIPTION
**Pull request by luispadron**
Before this change the CI and cache configs were being generated by `.github/workflows/xcode_select.sh` which makes it difficult for a contributor to find these and to then apply them.

Changes:
 - New Bazel configurations that make it easier to reproduce CI setup and debug issues locally.
 - New configs for testing specific tests that were previously only in the workflow file.
 - Removes the `user.bazelrc` generation CI was doing in favor of the new `ci` and `cache` configs.
 - Updates the workflows to use the new configs where needed.

Depends on #716

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -141,19 +165,26 @@ jobs:
...
       - uses: actions/checkout@v3
       - name: Select Xcode 14.2.0
         run: .github/workflows/xcode_select.sh
       - name: Run tests
-        run: bazelisk test tests/ios/lldb/app:objc_app_po_test  tests/ios/lldb/app:objc_app_variable_test --config lldb_ios_test
+        run: |
+          bazelisk test \
+            --config=ci \
+            --config=ios_lldb_test \
+            tests/ios/lldb/app:objc_app_po_test \
```
Comment (jszumski) [inline .github/workflows/tests.yml:179]:
```suggestion
            -- \
            tests/ios/lldb/app:objc_app_po_test \
```

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -162,4 +193,8 @@ jobs:
       - name: Select Xcode
         run: .github/workflows/xcode_select.sh
       - name: Build App
-        run: bazelisk build -s tests/ios/app/App --apple_platform_type=ios --ios_minimum_os=10.2  --ios_multi_cpus=i386,x86_64
+        run: |
+          bazelisk build \
+            --config=ci \
+            --config=ios_multi_arch_test \
+            tests/ios/app/App
```
Comment (jszumski) [inline .github/workflows/tests.yml:200]:
```suggestion
            -- \
            tests/ios/app/App
```

**Review by jszumski — APPROVED** [review]
This was always a pain point for me, thank you!

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -67,18 +77,23 @@ jobs:
           path: bazel-testlogs
 
   lts_ios_integration_tests:
-    name: Build and Test ( Virtual Frameworks + LTS )
+    name: Bazel 5 - Build and Test (Virtual Frameworks)
```
Comment (mattrobmattrob) [inline .github/workflows/tests.yml:80]: If we're going to be updating these, I'd prefer it stay beside the `Virtual Frameworks` portion since it's the "modifiers" to the build. I think the prefix makes the build list hard to parse:
Screenshot 2023-05-10 at 9 29 18 AM
So I guess:
```suggestion
    name: Build and Test (Virtual Frameworks, Bazel 5)
```

Comment (luispadron) [reply .github/workflows/tests.yml:80]: That makes sense, went back and forth on it

**Comment by luispadron** [comment]
Someone with Admin access will need to update the required jobs for the branch, if we think the name changes are worth it

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -162,4 +202,9 @@ jobs:
       - name: Select Xcode
         run: .github/workflows/xcode_select.sh
       - name: Build App
-        run: bazelisk build -s tests/ios/app/App --apple_platform_type=ios --ios_minimum_os=10.2  --ios_multi_cpus=i386,x86_64
```
Comment (mattrobmattrob) [inline .github/workflows/tests.yml:165]: Was `-s` intentionally removed? What did it do?

**Review by mattrobmattrob — APPROVED** [review]
Mostly nits. Take it or leave it. This is a great improvement for folks working on the ruleset. 👏

**Comment by thiagohmcruz** [comment]
> Someone with Admin access will need to update the required jobs for the branch, if we think the name changes are worth it

Probably @justinseanmartin could help here? Where exactly we would have to update those? I've never done that myself 😓

**Comment by luispadron** [comment]
> > Someone with Admin access will need to update the required jobs for the branch, if we think the name changes are worth it
> 
> Probably @justinseanmartin could help here? Where exactly we would have to update those? I've never done that myself 😓

It under Settings > Branches > Branch protection > `master` > Required checks

File: .github/workflows/tests.yml (hunk 4 above)
Comment (luispadron) [reply .github/workflows/tests.yml:165]: That prints the subcommands, i meant to add this `config:ci`. Do we think thats useful, it could be very verbose but its for CI so?

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -67,18 +80,24 @@ jobs:
...
-    name: Build and Test ( Virtual Frameworks + LTS )
+    name: Build and Test (Virtual Frameworks, Bazel 5)
     runs-on: macos-12
+    env:
+      USE_BAZEL_VERSION: 5.3.2
     steps:
       - uses: actions/checkout@v3
       - name: Select Xcode
         run: .github/workflows/xcode_select.sh
       - name: Build and Test
```
[the remark this answers is not in the document]
Comment (luispadron) [reply .github/workflows/tests.yml:91]: For sure, this confused me when i saw it was just running `build`

File: .bazelrc [hunk 6]
```diff
@@ -28,14 +28,48 @@ build --test_env=PATH=/usr/bin:/bin:/usr/sbin:/sbin
...
-
 # Delete the VM test suite for github
 build --deleted_packages tests/ios/vmd
 
 # Allow for server to start (GitHub actions can timeout)
 startup --local_startup_timeout_secs=240
 
+# - iOS Configuration
+# Required for tests/ios/... targets
+build:ios --apple_platform_type=ios
```
[the remark this answers is not in the document]
Comment (luispadron) [reply .bazelrc:39]: I prefer the shorter name in this case but if we think platform is more clear thats fine too!

File: .github/workflows/tests.yml (hunk 4 above)
Comment (luispadron) [reply .github/workflows/tests.yml:165]: Added this to `config=ci` as well as `--verbose_failures` figure its better to be verbose in CI.

File: .bazelrc (hunk 6 above)
[the remark this answers is not in the document]
Comment (mattrobmattrob) [reply .bazelrc:39]: All good then. 👍

**Comment by luispadron** [comment]
Weird @thiagohmcruz seeing a similar plist diff now:

```sh
-				INFOPLIST_FILE = ../../../tests/ios/frameworks/dynamic/a/Info.plist;
```

Let me double check I didnt change any underlying flags but its odd

**Comment by thiagohmcruz** [comment]
> 

Yeah, the question is if we even want that plist there since we set it to empty for apps and extensions we might as well be ok with setting it to empty here as well. The things is that removing it causes other issues apparently, I was trying to iterate on that [here](https://github.com/bazel-ios/rules_ios/pull/717)

**Comment by luispadron** [comment]
@thiagohmcruz Will wait for that PR then!

File: .github/workflows/tests.yml [hunk 7]
```diff
@@ -21,44 +21,60 @@ jobs:
...
         run: |
-          # Host config
-          bazelisk test --local_test_jobs=1 -- //... -//tests/ios/...
+          # non-iOS tests
+          bazelisk test \
+            --config=ci \
+            -- \
+            //... \
+            -//tests/ios/...
```
Comment (qyang-nj) [inline .github/workflows/tests.yml:29]: nit: I like this to be on a single line, so that it's easier to copy it into to terminal. Same as other commands as well.

**Review by qyang-nj — COMMENTED** [review]
Thanks! This hidden `user.bazlerc` bite me before.

**Comment by luispadron** [comment]
@thiagohmcruz fixed the Xcode tests, realized the ci generated rc file set the spawn strategy by default. I've enabled it by default in the main rc file now.

**Comment by luispadron** [comment]
This is ready to merge as soon as we can disable the required jobs that have been renamed. 

Once this merges we can enable the new ones. 

@thiagohmcruz @justinseanmartin whenever y'all have time or if you want to make me admin I can do it

**Review by mattrobmattrob — APPROVED** [review]

**Comment by luispadron** [comment]
Actually moving the renaming to a separate PR to not block this further: #722
